### ISSUE TITLE
cpuset not working with present arch linux (maybe cgroupv2?)

### ISSUE THREAD

**Issue by ebennett1980** [lpechacek/cpuset#40]
Problem seems to afflict arch linux on two different kernels, I think it has to do with them using control group v2 and cpuset expecting v1 perhaps? Arch upgraded in a recent update.

symptom looks like this;
root@monolith:~# cset shield --cpu=0-7
mount: /cpusets: none already mounted on /sys/fs/bpf.
cset: **> mount of cpuset filesystem failed, do you have permission?

Two kernels confirmed affected;

Linux monolith 5.4.85-1-vfio-lts #1 SMP Wed, 23 Dec 2020 06:46:51 +0000 x86_64 GNU/Linux
Linux magister 5.9.11-xanmod1-1 #1 SMP PREEMPT Tue, 01 Dec 2020 12:38:55 +0000 x86_64 GNU/Linux

**Comment by Werkov** [comment]
Hello.
What does the following say on your system:
```
cat /sys/fs/cgroup/cgroup.controllers
grep -E "cpuset|cgroup2" /proc/mounts
```
?

Also, what systemd version is this?

(I'm just checking, but I'd generally not expect cpuset working with v2 cpuset controller.)

**Comment by ebennett1980** [comment]
Results on the metal
cgroup2 /sys/fs/cgroup cgroup2 rw,nosuid,nodev,noexec,relatime 0 0
cpuset cpu io memory hugetlb pids rdma

Results in a VM
cgroup2 /sys/fs/cgroup cgroup2 rw,nosuid,nodev,noexec,relatime 0 0
cpuset cpu io memory hugetlb pids rdma

systemd is 248-5 on both

**Comment by Werkov** [comment]
Thanks. So the cpuset controller is bound to the v2 hierarchy and it also shows that systemd runs in the unified mode (that was perhaps the critical change between updates). I'm afraid the cset utility can't serve you in such a setup.

(As a workaround, you may switch via kernel cmdline back to the hybrid setup `systemd.unified_cgroup_hierarchy=0` and use cset as previously, or migrate your configuration to systemd cpuset implementation (I didn't test it).)

**Comment by Hubro** [comment]
@Werkov Do you have any advice on how to shield CPU cores now, with the new unified cgroup hierarchy? Does systemd have any functionality for this? I am unable to find any information about this.

**Comment by Werkov** [comment]
@Hubro With the recent systemd version you should be able to specify `AllowedCPUs=` directive. You can't set it directly on root (`-.slice`) but all 1st level children instead (`init.scope`, `system.slice` and `user.slice` in default setups). That way you can move the userspace tasks out of the way. (Note that kernel threads can still run on "shielded" CPUs but that is no different to `cset shield --kthread=off`(the default).)

**Comment by Hubro** [comment]
@Werkov Is it possible to set `AllowedCPUs` for already running slices without restarting them? Also, is there any way to inform the kernel which cores I want it to keep its threads on? In this case a kernel command line argument would be fine.

**Comment by Werkov** [comment]
Setting the cgroup attributes should work at runtime exactly as you did (alternatively you can edit slice unit or drop-in files and call `systemctl daemon-reload`), restart of slices is not necessary. (One possible catch with the runtime update is that NUMA memory won't be migrated with the change.) And you need the unified hierarchy for this to work with systemd (otherwise you'd have used `cset`, right?).

If you need "silence" on the cpu, then see [isolcpus](https://www.kernel.org/doc/html/latest/admin-guide/kernel-parameters.html?highlight=isolcpus) kernel cmdline. I'm just curious, why you need to keep kernel threads off your selected cores (is that due to RT constraints)?

**Comment by Hubro** [comment]
@Werkov My use case is a high performance virtual machine doing realtime tasks, so I'm doing everything I can to reduce latency and stutters on those cores.

I'm not entirely sure how isolcpus works. Will this keep kernel threads off those cores, or will it keep *everything* off them? I want to be able to do compilation and encoding tasks using all my host cores when my VMs are not running, so any kernel cmdline parameters that prevent that is not an option for me.

**Comment by fweisbec** [comment]
> Does that mean that cpusets can do _all_ the things that isolcpus can do? thinking

Not exactly, isolcpus is often used to disable scheduler load balancing on a CPU and that's the only
part where cpusets can help in a similar fashion (through cpuset.sched_load_balance), that and also telling
which task is allowed to go on a given set of CPUs. But that's where the similarity ends.

In fact isolcpus does much more as it also isolates from unbound kernel threads, workqueues, timers, etc...

You can also try "nohz_full=". It will isolate your CPU pretty much as well as isolcpus does and it will also further
deactivate the timer tick on the host CPU, avoiding interrupting the guest vCPUs. If you combine that with cpusets to move all unrelated tasks from the CPUs running the guests, you might have good results. Oh and don't forget to re-affine interrupts out
of the CPUs running guest as well (https://www.kernel.org/doc/Documentation/IRQ-affinity.txt).
I'm writing a series of articles about that if that can help: https://www.suse.com/c/cpu-isolation-introduction-part-1/

**Comment by Werkov** [comment]
>     3. With the introduction of SystemD in SUSE products I heard horror stories about how SystemD freezes when external programs manipulate its group setting. It was LTP at that time taking SystemD down in product testing.

Fortunately, this is irrelevant for cgroup hierarchies that are not managed by systemd. Practically, the cpuset utility could be used safely with systemd prior v244 (which introduced support for cpuset in systemd).

> At this point, I'd like to know your opinion(s) about whether it is safe to manipulate systems cgroup settings without the daemon's consent. 

Nowadays with v2, it is safe when the operations are carried out in a dedicated subtree only. Full description is in the document about [cgroup delegation with systemd](https://systemd.io/CGROUP_DELEGATION/).

**Comment by haelix888** [comment]
Has anyone managed to set isolcpus on Arch using EFISTUB (efibootmgr) by any chance?
The kernel parameter is not getting picked up.

**Comment by fweisbec** [comment]
> Has anyone managed to set isolcpus on Arch using EFISTUB (efibootmgr) by any chance?
> The kernel parameter is not getting picked up.
> (linux-lts 5.10.40)
> 
> Edit: possibly related: https://lore.kernel.org/lkml/20200414215715.GB182757@xz-x1/T/#u

No idea but you can still hardcode kernel boot option with CONFIG_CMDLINE_BOOL + CONFIG_CMDLINE

**Comment by haelix888** [comment]
My issue was with the BIOS. It somehow deduplicates boot entries having the same image, but different parameters. In my case I can get it to work by entering bios and modifying boot priority order (even if efibootmgr correctly reports the order).

**Comment by larseggert** [comment]
I did reinstall grub with `GRUB_CMDLINE_LINUX='systemd.unified_cgroup_hierarchy=false'`, but I still cannot use `cset`. Is this working for anyone?

```
lars@t-linux64-ms-280:~$ sudo cset set --cpu=0-1,4-5 --set=system --cpu_exclusive
mount: /cpusets: unknown filesystem type 'cpuset'.
       dmesg(1) may have more information after failed mount system call.
cset: **> mount of cpuset filesystem failed, do you have permission?
```

**Comment by Werkov** [comment]
What's your kernel's `CONFIG_CPUSETS_V1`? (config files are e.g. in `/proc/config.gz` or under `/boot`)  You'd [need that](https://github.com/SUSE/cpuset/issues/49#issuecomment-2498562881) too.

**Comment by larseggert** [comment]
It's a stock Ubuntu 24.04 kernel. (Can't access the machine right now, will respond d with details next week.)

**Comment by larseggert** [comment]
`/boot/config-6.14.0-27-generic` has
```
# CONFIG_CPUSETS_V1 is not set
```
Older kernel configs don't have that commented line, but don't set it either.

**Comment by Werkov** [comment]
Thanks for chekcing, so you shouldn't need `systemd.unified_cgroup_hierarchy=false` but you need a kernel which enables `CONFIG_CPUSETS_V1` (it's a new config option to deprecate v1)).

Which scenario of cpuset utility (it only supports cpuset v1) do you need? Shielding?

**Comment by larseggert** [comment]
I've tried the previous stock 6.11 Ubuntu kernel, and that seems to still work, so I reverted to that for now.

Yes, shielding, although via `cset set` and `cset proc` for some customization.
